src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs

```diff
@@ -0,0 +1,287 @@
+
+            var endpoint = endpointDocument?.ToEndpoint();
+            if (endpoint != null)
+            {
+                var timeSeries = await session
+                    .IncrementalTimeSeriesFor(documentId, ThroughputTimeSeriesName)
+                    .GetAsync(token: cancellationToken);
+
+                endpoint.LastCollectedDate = DateOnly.FromDateTime(timeSeries.LastOrDefault()?.Timestamp ?? DateTime.MinValue);
+            }
+
+            endpoints.Add((id, endpoint));
+        }
+
+        return endpoints;
+    }
+
+    public async Task SaveEndpoint(Endpoint endpoint, CancellationToken cancellationToken)
+    {
+        var document = endpoint.ToEndpointDocument();
+
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        await session.StoreAsync(document, document.GenerateDocumentId(), cancellationToken);
+        await session.SaveChangesAsync(cancellationToken);
+    }
+
+    public async Task<IDictionary<string, IEnumerable<ThroughputData>>> GetEndpointThroughputByQueueName(IEnumerable<string> queueNames, CancellationToken cancellationToken)
+    {
+        var results = queueNames.ToDictionary(queueName => queueName, queueNames => new List<ThroughputData>() as IEnumerable<ThroughputData>);
+
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        var query = session.Query<EndpointDocument>()
+            .Where(document => document.SanitizedName.In(queueNames))
+            .Include(builder => builder.IncludeTimeSeries(ThroughputTimeSeriesName));
+
+        var documents = await query.ToListAsync(cancellationToken);
+
+        foreach (var document in documents)
+        {
+            var timeSeries = await session
+                .IncrementalTimeSeriesFor(document.GenerateDocumentId(), ThroughputTimeSeriesName)
+                .GetAsync(token: cancellationToken);
+
+            if (results.TryGetValue(document.SanitizedName, out var throughputDatas) &&
+                throughputDatas is List<ThroughputData> throughputDataList)
+            {
+                var endpointDailyThroughputs = timeSeries.Select(entry => new EndpointDailyThroughput(DateOnly.FromDateTime(entry.Timestamp), (long)entry.Value));
+                var throughputData = new ThroughputData(endpointDailyThroughputs)
+                {
+                    ThroughputSource = document.EndpointId.ThroughputSource
+                };
+                throughputDataList.Add(throughputData);
+            }
+        }
+
+        return results;
+    }
+
+    public async Task RecordEndpointThroughput(string endpointName, ThroughputSource throughputSource, IEnumerable<EndpointDailyThroughput> throughput, CancellationToken cancellationToken)
+    {
+        if (!throughput.Any())
```

Comment:
Possible multiple enumeration of  `throughput` ?

Reply:
Fixed